### PR TITLE
test/vault: use real routing for the integration test, with multiple vaults

### PR DESCRIPTION
**Pull request by maqi**
This PR contains the work to make vault use real routing (with mock_parsec) during the integration test.
The testing environment contains a network of vaults, and the client requests are handled by the network, instead of just one vault.

### CODE REVIEW

File: src/config_handler.rs [hunk 1]
```diff
@@ -185,7 +185,7 @@ impl Config {
         } else if arg == ARGS[11] {
             self.network_config.our_type = unwrap!(value.parse());
         } else {
-            #[cfg(not(feature = "mock"))]
+            #[cfg(not(any(feature = "mock", feature = "mock-parsec")))]
```
Comment (jeanphilippeD) [inline src/config_handler.rs]: seem like it would be better have a `mock_base` feature `["routing/mock_base", "fxhash", "hex_fmt"]
And only config on this.

Then
mock = [mock_base]
mock_parsec = ["routing/mock_parsec", "routing/mock_crypto", "mock_base", "phase-one"]

Then no need to have these more complicated feature gate.

File: Cargo.toml [hunk 2]
```diff
@@ -47,4 +48,5 @@ doc = false
 
 [features]
 mock = ["routing/mock_base", "fxhash", "hex_fmt"]
+mock-parsec = ["routing/mock_parsec", "routing/mock_crypto", "fxhash", "hex_fmt", "phase-one"]
 phase-one = []
```
Comment (jeanphilippeD) [inline Cargo.toml]: phase_one?
mock_parsec?

for consistency with routing at least?

File: src/vault.rs [hunk 3]
```diff
@@ -188,8 +196,18 @@ impl Vault {
...
+        let mut max_poll = 0;
         loop {
+            max_poll += 1;
+            if processed || max_poll > 1000 {
+                return processed;
+            }
+
+            #[cfg(feature = "mock-parsec")]
+            {
+                processed = self.routing_node.borrow_mut().poll();
```
Comment (jeanphilippeD) [inline src/vault.rs]: Why is this needed, why is `handle_selected_operation`, and should we make it sufficient somehow if not?

File: src/vault.rs [hunk 4]
```diff
@@ -188,8 +196,18 @@ impl Vault {
     /// Returns whether at least one event was processed.
     pub fn poll(&mut self) -> bool {
         let mut processed = false;
-
+        let mut max_poll = 0;
         loop {
+            max_poll += 1;
+            if processed || max_poll > 1000 {
```
Comment (jeanphilippeD) [inline src/vault.rs]: do we really need that within that function, is poll not call in a loop itself?

File: src/vault.rs [hunk 5]
```diff
@@ -219,12 +237,12 @@ impl Vault {
...
                     }
                     idx => {
-                        if let Err(err) = self
+                        if let Err(_err) = self
                             .routing_node
                             .borrow_mut()
                             .handle_selected_operation(idx)
                         {
-                            warn!("Could not process operation: {}", err);
+                            // warn!("Could not process operation: {}", err);
```
Comment (jeanphilippeD) [inline src/vault.rs]: why?

File: Cargo.toml (hunk 2 above)
Comment (maqi) [reply Cargo.toml]: yeah, good suggestion, I will change that

File: tests/common/mod.rs [hunk 6]
```diff
@@ -76,6 +86,43 @@ impl Environment {
...
+
+        let network = Network::new(Default::default());
+
+        let mut env = Self {
+            rng: rng::new(),
+            network,
+            vaults: Default::default(),
+        };
+
+        if num_vaults > 1 {
```
Comment (jeanphilippeD) [inline tests/common/mod.rs]: only 1 vault with this PR? so can remove that part of the if?

File: src/config_handler.rs (hunk 1 above)
Comment (maqi) [reply src/config_handler.rs]: yeah, I will try to change

File: tests/common/mod.rs [hunk 7]
```diff
@@ -93,6 +141,20 @@ impl Environment {
...
+    // Poll the mock network and the environment's vault.
+    pub fn poll(&mut self) {
+        let mut max_poll = 0;
+        while max_poll < 100 {
+            max_poll += 1;
+            self.network.poll();
+            let _ = self.vaults.iter_mut().for_each(|vault| {
+                let _ = vault.inner.poll();
+            });
+            FakeClock::advance_time(1001);
```
Comment (jeanphilippeD) [inline tests/common/mod.rs]: I see why we need that `FakeClock::advance_time(1001);`, I'm not clear why it need to ignore the progress part?
It seem a bit strange to always move the test forward 100 seconds each time we call poll

File: tests/common/mod.rs [hunk 8]
```diff
@@ -154,9 +218,54 @@ impl TestVault {
...
+
+        Self {
+            inner,
+            _root_dir: root_dir,
+            _command_tx: command_tx,
+        }
+    }
+
+    #[cfg(feature = "mock-parsec")]
+    fn new_with_real_routing_first() -> Self {
```
Comment (jeanphilippeD) [inline tests/common/mod.rs]: Seem to be mostly duplicate with new_with_real_routing

File: src/vault.rs (hunk 3 above)
Comment (maqi) [reply src/vault.rs]: only `poll` checks the timer advancing, `handle_selected_operation` doesnt

File: tests/common/mod.rs (hunk 8 above)
Comment (maqi) [reply tests/common/mod.rs]: I will carry out a de-depulication

File: src/vault.rs (hunk 3 above)
Comment (jeanphilippeD) [reply src/vault.rs]: What do you mean by check the timer advancing?
And why is handle_selected_operation not, should it?

File: tests/common/mod.rs (hunk 7 above)
Comment (maqi) [reply tests/common/mod.rs]: will have a double check if this still applies with the other changes.

File: tests/common/mod.rs (hunk 6 above)
Comment (maqi) [reply tests/common/mod.rs]: there are multiple vaults.
I think we will support multiple vaults with real routing soon, so don't want to change backwards and forwards.
The existing code already take multple vaults with mock_routing. So just want to ensure the code behave the same for both.

File: src/vault.rs (hunk 5 above)
Comment (maqi) [reply src/vault.rs]: there are too many logs for this so got it disabled.
I think it is due to the `try_ready` method used

File: src/vault.rs (hunk 4 above)
Comment (maqi) [reply src/vault.rs]: I remember I tried, it doesn't work. will have a double check again.

File: tests/common/mod.rs (hunk 6 above)
Comment (jeanphilippeD) [reply tests/common/mod.rs]: It does not make sense in the context of this PR, it would be best to either complete the work so it work with multiple vault, or remove the code that is not get us to fully functional multiple vaults.

Comment (maqi) [reply tests/common/mod.rs]: I thought PR 903 (https://github.com/maidsafe/safe_vault/pull/903) is to deal with `purely one vault network`

Yeah, probably this PR shall be better work with multiple vaults.
I will add a WIP label then.

File: src/vault.rs (hunk 3 above)
Comment (maqi) [reply src/vault.rs]: the `routing::Node::poll` eventually calls `routing::StateMachine::try_step`, which checks the timed_out_tokens
However, the `handle_selected_operation ` doesn't, as it is just for mock_routing.

File: Cargo.toml [hunk 9]
```diff
@@ -38,13 +40,13 @@ unwrap = "~1.2.1"
...
 [[bin]]
 name = "safe_vault"
 doc = false
 
 [features]
-mock = ["routing/mock_base", "fxhash", "hex_fmt"]
-phase-one = []
+mock_base = ["routing/mock_base", "fxhash", "hex_fmt"]
+mock = ["mock_base"]
+mock_parsec = ["routing/mock_parsec", "routing/mock_crypto", "mock_base"]
```
Comment (octol) [inline Cargo.toml]: So `mock` means `mock_routing`, yet `mock_parsec` depends on `mock` but still uses real routing?

File: src/vault.rs [hunk 10]
```diff
@@ -188,10 +208,15 @@ impl Vault {
 
     /// Processes any outstanding network events and returns. Does not block.
     /// Returns whether at least one event was processed.
+    #[allow(unused_assignments)]
```
Comment (octol) [inline src/vault.rs]: Maybe mark the variable `_processed` to limit the scope of this rule?

File: tests/common/mod.rs [hunk 11]
```diff
@@ -128,15 +194,20 @@ trait AsMutSlice<T> {
...
 impl TestVault {
     /// Create a test Vault within a group.
-    fn new(consensus_group: Option<ConsensusGroupRef>, mut rng: TestRng) -> Self {
+    #[cfg(feature = "mock")]
+    fn new_with_mock_routing(
+        consensus_group: Option<ConsensusGroupRef>,
+        network: &Network,
+    ) -> Self {
         let root_dir = unwrap!(TempDir::new("safe_vault"));
+        trace!("creating a test vault at root_dir {:?}", root_dir);
```
Comment (octol) [inline tests/common/mod.rs]: s/creating/Creating ?

File: src/vault.rs [hunk 12]
```diff
@@ -189,9 +209,13 @@ impl Vault {
     /// Processes any outstanding network events and returns. Does not block.
     /// Returns whether at least one event was processed.
     pub fn poll(&mut self) -> bool {
-        let mut processed = false;
-
+        let mut _processed = false;
```
Comment (jeanphilippeD) [inline src/vault.rs]: why the _

Comment (maqi) [reply src/vault.rs]: just trying to limit the scope of rule, as mentioned at: https://github.com/maidsafe/safe_vault/pull/910#discussion_r359381896

Without `_`, we will have to use `#[allow(unused_assignments)]` above the function

Comment (jeanphilippeD) [reply src/vault.rs]: I don't get it, we return the value.
Your problem is that you probably should just have:

`let mut processed;`

Comment (maqi) [reply src/vault.rs]: I tried `let mut processed;`
there is a complain of `value assigned to processed is never read`
I think it's because the `returning` is not considered as a `read`, as there is an initial value assigned, however may not be used before being set to `true` ?

File: src/client_handler.rs [hunk 13]
```diff
@@ -1555,7 +1582,7 @@ impl ClientHandler {
         let valid = if let Some(signature) = signature {
             self.is_valid_client_signature(public_id, request, &message_id, &signature)
         } else {
-            warn!(
+            println!(
```
Comment (nbaksalyar) [inline src/client_handler.rs]: Is there a reason to change from one of the log macros to `println!`?

File: src/vault.rs [hunk 14]
```diff
@@ -189,9 +209,13 @@ impl Vault {
...
         loop {
+            #[cfg(feature = "mock_parsec")]
```
Comment (nbaksalyar) [inline src/vault.rs]: Probably something like
```rust
if cfg!(feature = "mock_parsec") {
  ...
}
```
would be more idiomatic?

File: tests/common/mod.rs [hunk 15]
```diff
@@ -96,6 +142,20 @@ impl Environment {
...
+    // Poll the mock network and the environment's vault.
+    pub fn poll(&mut self) {
+        let mut processed = true;
+        while processed {
+            processed = false;
+            self.network.poll();
+            self.vaults
+                .iter_mut()
+                .for_each(|vault| processed = processed || vault.inner.poll());
+            FakeClock::advance_time(1001);
```
Comment (nbaksalyar) [inline tests/common/mod.rs]: `1001` looks like an arbitrary magic number, so maybe put it into a constant or add a comment?

File: tests/common/mod.rs [hunk 16]
```diff
@@ -220,23 +333,27 @@ pub trait TestClientTrait {
...
-            Ok(Event::ConnectedTo {
-                peer: Peer::Node { ref node_info },
-            }) if node_info == conn_info => (),
-            x => unexpected!(x),
+        loop {
+            match self.rx().try_recv() {
+                Ok(Event::ConnectedTo {
+                    peer: Peer::Node { ref node_info },
+                }) if node_info == conn_info => return,
+                // _ => (),
```
Comment (nbaksalyar) [inline tests/common/mod.rs]: Looks like dead code

File: src/client_handler.rs (hunk 13 above)
Comment (maqi) [reply src/client_handler.rs]: Ah, missed this one. a left over from the debuggings. will change it back.

File: tests/common/mod.rs (hunk 15 above)
Comment (maqi) [reply tests/common/mod.rs]: just used the same as in routing. will put a comment in

File: tests/common/mod.rs (hunk 16 above)
Comment (maqi) [reply tests/common/mod.rs]: forgot to remove, will do

File: src/vault.rs (hunk 14 above)
Comment (maqi) [reply src/vault.rs]: Cannot do that.

Although the code won't be executed when compile with `mock`, the compiler will still looking for routing_node having a `poll` function, which is not provided (needed) by mock_routing.
